File: koan_ansi/convert.py
"""Core matcher: image → grid of (glyph, fg, bg) cells.

Model: each character cell is sampled as 2×2 quadrant mean colors [TL,TR,BL,BR].
Every candidate (glyph, fg, bg) reconstructs those four quadrants as
``fg*cov + bg*(1-cov)``; we pick the candidate minimizing channel-weighted
squared error plus, for shade glyphs, a chroma-noise penalty — a shade mixing
two hue-distant colors (green dots over blue) matches the *average* but reads
as speckle, so hue-clashing fg/bg pairs are penalized (opponent space, luma
removed). NOISE_K value inherited from koan-site's tuned converter.

The brute force is exact: all candidates are scored, vectorized as one matrix
product per block of cells (error = const(cand) - 2·q·recon, the q² term is
constant per cell and drops out of the argmin).
"""

from __future__ import annotations

import numpy as np
from PIL import Image

from .charset import CHARSETS, Glyph
from .palette import CHANNEL_W, DEFAULT_NOISE_K, OP_DIST2, PALETTE


def auto_rows(width: int, height: int, cols: int) -> int:
    """Character cells are 1:2 (w:h): rows = cols * aspect / 2."""
    return max(1, round(cols * (height / width) / 2))


def quad_grid(img: Image.Image, cols: int, rows: int) -> np.ndarray:
    """(rows, cols, 4, 3) float32 quadrant mean colors, order [TL, TR, BL, BR]."""
    rgb = img.convert("RGB")
    w2, h2 = cols * 2, rows * 2
    # BOX = true area average when downscaling; BICUBIC avoids blockiness when
    # a small source is scaled up to the quadrant grid.
    resample = (
        Image.Resampling.BOX
        if rgb.width >= w2 and rgb.height >= h2
        else Image.Resampling.BICUBIC
    )
    small = rgb.resize((w2, h2), resample)
    a = np.asarray(small, dtype=np.float32)
    return a.reshape(rows, 2, cols, 2, 3).transpose(0, 2, 1, 3, 4).reshape(rows, cols, 4, 3)


def build_candidates(
    charset: list[Glyph], noise_k: float, include_shades: bool = True
) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """All (glyph, fg, bg) candidates over the 16-color palette.

    Returns (rw, const, meta): weighted recon vectors (N,12), per-candidate
    constant term Σw·r² + noise penalty (N,), and meta int32 (N,3) rows of
    [glyph_index, fg, bg]. Glyph indices refer to the FULL charset list even
    when shades are excluded.
    """
    recon_list: list[np.ndarray] = []
    meta: list[tuple[int, int, int]] = []
    pen: list[float] = []
    for gi, g in enumerate(charset):
        if g.shade is not None and not include_shades:
            continue
        cov = np.asarray(g.cov, dtype=np.float32)[:, None]  # (4,1)
        mix = g.shade * (1.0 - g.shade) if g.shade is not None else 0.0
        fgs = (0,) if g.cov == (0, 0, 0, 0) else tuple(range(16))
        for fg in fgs:
            bgs = (0,) if g.cov == (1, 1, 1, 1) else tuple(range(16))
            for bg in bgs:
                recon_list.append(PALETTE[fg] * cov + PALETTE[bg] * (1.0 - cov))
                meta.append((gi, fg, bg))
                pen.append(noise_k * mix * 4.0 * float(OP_DIST2[fg, bg]))
    recon = np.stack(recon_list)  # (N,4,3)
    meta_a = np.asarray(meta, dtype=np.int32)
    const = (recon * recon * CHANNEL_W).sum(axis=(1, 2)) + np.asarray(pen, dtype=np.float32)
    rw = (recon * CHANNEL_W).reshape(len(meta), 12).astype(np.float32)
    return rw, const.astype(np.float32), meta_a


def match(
    quads: np.ndarray,
    charset: list[Glyph],
    noise_k: float = DEFAULT_NOISE_K,
    include_shades: bool = True,
    chunk: int = 16384,
) -> np.ndarray:
    """Best candidate per cell → packed int32 grid (glyph<<8)|(fg<<4)|bg."""
    rows, cols = quads.shape[:2]
    rw, const, meta = build_candidates(charset, noise_k, include_shades)
    q = quads.reshape(-1, 12).astype(np.float32)
    best = np.empty(q.shape[0], dtype=np.int64)
    for i in range(0, q.shape[0], chunk):
        score = const[None, :] - 2.0 * (q[i : i + chunk] @ rw.T)
        best[i : i + chunk] = score.argmin(axis=1)
    m = meta[best]
    glyph = m[:, 0].reshape(rows, cols)
    fg = m[:, 1].reshape(rows, cols)
    bg = m[:, 2].reshape(rows, cols)
    return ((glyph << 8) | (fg << 4) | bg).astype(np.int32)


def unpack(packed: np.ndarray) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """packed → (glyph, fg, bg). Glyph is the high bits (no mask: extended
    charset indices exceed 4 bits; fg/bg are always 4-bit palette indices)."""
    return packed >> 8, (packed >> 4) & 0xF, packed & 0xF


def fs_dither_quads(quads: np.ndarray) -> np.ndarray:
    """Serpentine Floyd–Steinberg on the quadrant image, snapped to the
    16-color palette (channel-weighted nearest). Use with include_shades=False:
    FS already breaks gradients into structure, shades would fight it."""
    rows, cols = quads.shape[:2]
    h2, w2 = rows * 2, cols * 2
    img = (
        quads.reshape(rows, cols, 2, 2, 3)
        .transpose(0, 2, 1, 3, 4)
        .reshape(h2, w2, 3)
        .astype(np.float32)
        .copy()
    )
    out = np.empty_like(img)
    for y in range(h2):
        serp = (y & 1) == 1
        xs = range(w2 - 1, -1, -1) if serp else range(w2)
        for x in xs:
            c = np.clip(img[y, x], 0.0, 255.0)
            i = int((((PALETTE - c) ** 2) * CHANNEL_W).sum(axis=1).argmin())
            out[y, x] = PALETTE[i]
            e = c - PALETTE[i]
            xn = x - 1 if serp else x + 1  # next pixel in scan direction
            xp = x + 1 if serp else x - 1
            if 0 <= xn < w2:
                img[y, xn] += e * (7 / 16)
            if y + 1 < h2:
                img[y + 1, x] += e * (5 / 16)
                if 0 <= xp < w2:
                    img[y + 1, xp] += e * (3 / 16)
                if 0 <= xn < w2:
                    img[y + 1, xn] += e * (1 / 16)
    return (
        out.reshape(rows, 2, cols, 2, 3).transpose(0, 2, 1, 3, 4).reshape(rows, cols, 4, 3)
    )


def match_truecolor(
    quads: np.ndarray, charset: list[Glyph]
) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """Exact-color mode: spatial glyphs only. A flat cell already hits its
    target color exactly, so shade glyphs add nothing but texture (█ is the
    same dropped duplicate of space). Returns (glyph, fg_rgb u8, bg_rgb u8)."""
    rows, cols = quads.shape[:2]
    cand = [
        (i, g)
        for i, g in enumerate(charset)
        if g.shade is None and g.cov != (1, 1, 1, 1)
    ]
    errs, fgs, bgs, gidx = [], [], [], []
    for i, g in cand:
        cov = np.asarray(g.cov, dtype=bool)
        nf = int(cov.sum())
        if nf == 0:
            bg = quads.mean(axis=2)
            fg = bg
        else:
            fg = quads[:, :, cov, :].mean(axis=2)
            bg = quads[:, :, ~cov, :].mean(axis=2) if nf < 4 else fg
        recon = np.where(cov[None, None, :, None], fg[:, :, None, :], bg[:, :, None, :])
        errs.append((((quads - recon) ** 2) * CHANNEL_W).sum(axis=(2, 3)))
        fgs.append(fg)
        bgs.append(bg)
        gidx.append(i)
    E = np.stack(errs)  # (G, rows, cols)
    k = E.argmin(axis=0)
    glyph = np.asarray(gidx, dtype=np.int32)[k]
    FG = np.stack(fgs)
    BG = np.stack(bgs)
    ii, jj = np.meshgrid(np.arange(rows), np.arange(cols), indexing="ij")
    fg = np.clip(np.rint(FG[k, ii, jj]), 0, 255).astype(np.uint8)
    bg = np.clip(np.rint(BG[k, ii, jj]), 0, 255).astype(np.uint8)
    return glyph, fg, bg


def convert_image(
    img: Image.Image | str,
    cols: int = 120,
    rows: int = 0,
    charset: str = "classic",
    dither: str = "none",
    noise_k: float = DEFAULT_NOISE_K,
) -> np.ndarray:
    """High-level one-shot: image (path or PIL) → packed vga16 grid."""
    if isinstance(img, str):
        img = Image.open(img)
    if not rows:
        rows = auto_rows(img.width, img.height, cols)
    quads = quad_grid(img, cols, rows)
    cs = CHARSETS[charset]
    if dither == "fs":
        return match(fs_dither_quads(quads), cs, noise_k, include_shades=False)
    return match(quads, cs, noise_k)

File: web/src/convert.ts
// Core matcher — TS port of ansiconv/convert.py (same math, same NOISE_K trick).
// Error = channel-weighted squared RGB over the cell's 4 quadrants
//       + chroma-noise penalty for shade glyphs mixing hue-distant pairs.
// Scored as konst(cand) - 2·(q·reconW): the q² term is per-cell constant.

import { type Glyph } from './charset'
import { CHANNEL_W, DEFAULT_NOISE_K, OP_DIST2, PALETTE } from './palette'

export function autoRows(w: number, h: number, cols: number): number {
  return Math.max(1, Math.round((cols * (h / w)) / 2))
}

/** Image → quadrant colors, cell-major Float32Array [cell][quad TL,TR,BL,BR][rgb]. */
export function quadsFromImage(
  src: CanvasImageSource & { width: number; height: number },
  cols: number,
  rows: number,
): Float32Array {
  const w2 = cols * 2
  const h2 = rows * 2
  const cv = document.createElement('canvas')
  cv.width = w2
  cv.height = h2
  const cx = cv.getContext('2d', { willReadFrequently: true })!
  cx.fillStyle = '#000' // transparent sources composite over black, like a terminal
  cx.fillRect(0, 0, w2, h2)
  cx.imageSmoothingEnabled = true
  cx.imageSmoothingQuality = 'high'
  cx.drawImage(src, 0, 0, w2, h2)
  const d = cx.getImageData(0, 0, w2, h2).data
  const q = new Float32Array(rows * cols * 12)
  for (let cy = 0; cy < rows; cy++)
    for (let cxi = 0; cxi < cols; cxi++) {
      const base = (cy * cols + cxi) * 12
      for (let k = 0; k < 4; k++) {
        const px = cxi * 2 + (k & 1)
        const py = cy * 2 + (k >> 1)
        const o = (py * w2 + px) * 4
        q[base + k * 3] = d[o]
        q[base + k * 3 + 1] = d[o + 1]
        q[base + k * 3 + 2] = d[o + 2]
      }
    }
  return q
}

export interface CandTable {
  reconW: Float32Array // N×12, premultiplied by channel weights
  konst: Float32Array // N — Σw·r² + noise penalty
  meta: Int32Array // N×3 — glyph index (in full charset), fg, bg
  n: number
}

export function buildCandidates(
  charset: Glyph[],
  noiseK: number = DEFAULT_NOISE_K,
  includeShades = true,
): CandTable {
  const reconW: number[] = []
  const konst: number[] = []
  const meta: number[] = []
  for (let gi = 0; gi < charset.length; gi++) {
    const g = charset[gi]
    if (g.shade !== null && !includeShades) continue
    const mix = g.shade !== null ? g.shade * (1 - g.shade) : 0
    const isSpace = g.cov.every((c) => c === 0)
    const isFull = g.cov.every((c) => c === 1)
    for (let fg = 0; fg < 16; fg++) {
      if (isSpace && fg > 0) break // space: fg irrelevant
      for (let bg = 0; bg < 16; bg++) {
        if (isFull && bg > 0) break // █: bg irrelevant
        let k0 = noiseK * mix * 4 * OP_DIST2[fg * 16 + bg]
        for (let k = 0; k < 4; k++) {
          const c = g.cov[k]
          for (let ch = 0; ch < 3; ch++) {
            const r = PALETTE[fg][ch] * c + PALETTE[bg][ch] * (1 - c)
            reconW.push(r * CHANNEL_W[ch])
            k0 += r * r * CHANNEL_W[ch]
          }
        }
        konst.push(k0)
        meta.push(gi, fg, bg)
      }
    }
  }
  return {
    reconW: Float32Array.from(reconW),
    konst: Float32Array.from(konst),
    meta: Int32Array.from(meta),
    n: konst.length,
  }
}

/** Best (glyph, fg, bg) per cell → packed Int32Array (glyph<<8)|(fg<<4)|bg. */
export function matchCells(quads: Float32Array, nCells: number, t: CandTable): Int32Array {
  const { reconW, konst, meta, n } = t
  const out = new Int32Array(nCells)
  for (let c = 0; c < nCells; c++) {
    const qb = c * 12
    let best = 0
    let bestScore = Infinity
    for (let i = 0; i < n; i++) {
      const rb = i * 12
      let dot = 0
      for (let j = 0; j < 12; j++) dot += quads[qb + j] * reconW[rb + j]
      const s = konst[i] - 2 * dot
      if (s < bestScore) {
        bestScore = s
        best = i
      }
    }
    out[c] = (meta[best * 3] << 8) | (meta[best * 3 + 1] << 4) | meta[best * 3 + 2]
  }
  return out
}

/** Serpentine Floyd–Steinberg on the quadrant grid, snapped to the 16-color
 * palette. Use with includeShades=false — FS already turns gradients into
 * structure, shade glyphs would fight it. */
export function fsDitherQuads(quads: Float32Array, cols: number, rows: number): Float32Array {
  const w2 = cols * 2
  const h2 = rows * 2
  // unpack cell-major quads to a scanline image
  const img = new Float32Array(w2 * h2 * 3)
  for (let cy = 0; cy < rows; cy++)
    for (let cx = 0; cx < cols; cx++)
      for (let k = 0; k < 4; k++) {
        const src = ((cy * cols + cx) * 4 + k) * 3
        const dst = ((cy * 2 + (k >> 1)) * w2 + cx * 2 + (k & 1)) * 3
        img[dst] = quads[src]
        img[dst + 1] = quads[src + 1]
        img[dst + 2] = quads[src + 2]
      }
  const out = new Float32Array(img.length)
  for (let y = 0; y < h2; y++) {
    const serp = (y & 1) === 1
    for (let s = 0; s < w2; s++) {
      const x = serp ? w2 - 1 - s : s
      const o = (y * w2 + x) * 3
      const r = Math.min(255, Math.max(0, img[o]))
      const g = Math.min(255, Math.max(0, img[o + 1]))
      const b = Math.min(255, Math.max(0, img[o + 2]))
      let bi = 0
      let bd = Infinity
      for (let p = 0; p < 16; p++) {
        const dr = r - PALETTE[p][0]
        const dg = g - PALETTE[p][1]
        const db = b - PALETTE[p][2]
        const d = dr * dr * CHANNEL_W[0] + dg * dg * CHANNEL_W[1] + db * db * CHANNEL_W[2]
        if (d < bd) {
          bd = d
          bi = p
        }
      }
      out[o] = PALETTE[bi][0]
      out[o + 1] = PALETTE[bi][1]
      out[o + 2] = PALETTE[bi][2]
      const er = r - PALETTE[bi][0]
      const eg = g - PALETTE[bi][1]
      const eb = b - PALETTE[bi][2]
      const xn = serp ? x - 1 : x + 1
      const xp = serp ? x + 1 : x - 1
      const spread = (ox: number, oy: number, w: number) => {
        if (ox < 0 || ox >= w2 || oy >= h2) return
        const t = (oy * w2 + ox) * 3
        img[t] += er * w
        img[t + 1] += eg * w
        img[t + 2] += eb * w
      }
      spread(xn, y, 7 / 16)
      spread(xp, y + 1, 3 / 16)
      spread(x, y + 1, 5 / 16)
      spread(xn, y + 1, 1 / 16)
    }
  }
  // repack to cell-major
  const q2 = new Float32Array(quads.length)
  for (let cy = 0; cy < rows; cy++)
    for (let cx = 0; cx < cols; cx++)
      for (let k = 0; k < 4; k++) {
        const dst = ((cy * cols + cx) * 4 + k) * 3
        const src = ((cy * 2 + (k >> 1)) * w2 + cx * 2 + (k & 1)) * 3
        q2[dst] = out[src]
        q2[dst + 1] = out[src + 1]
        q2[dst + 2] = out[src + 2]
      }
  return q2
}

export interface TrueResult {
  glyph: Int32Array
  fg: Uint8ClampedArray // n×3
  bg: Uint8ClampedArray // n×3
}

/** Truecolor mode: spatial glyphs only (flat cells already hit their color
 * exactly; shades would only add texture, █ duplicates space). */
export function matchTruecolor(quads: Float32Array, nCells: number, charset: Glyph[]): TrueResult {
  const cands = charset
    .map((g, i) => ({ g, i }))
    .filter(({ g }) => g.shade === null && !g.cov.every((c) => c === 1))
  const glyph = new Int32Array(nCells)
  const fg = new Uint8ClampedArray(nCells * 3)
  const bg = new Uint8ClampedArray(nCells * 3)
  for (let c = 0; c < nCells; c++) {
    const qb = c * 12
    let bestErr = Infinity
    let bg0 = 0, bg1 = 0, bg2 = 0, bf0 = 0, bf1 = 0, bf2 = 0, bgi = 0
    for (const { g, i } of cands) {
      let f0 = 0, f1 = 0, f2 = 0, b0 = 0, b1 = 0, b2 = 0, nf = 0
      for (let k = 0; k < 4; k++) {
        const r = quads[qb + k * 3]
        const gg = quads[qb + k * 3 + 1]
        const bb = quads[qb + k * 3 + 2]
        if (g.cov[k] === 1) {
          f0 += r; f1 += gg; f2 += bb; nf++
        } else {
          b0 += r; b1 += gg; b2 += bb
        }
      }
      if (nf > 0) {
        f0 /= nf; f1 /= nf; f2 /= nf
      }
      const nb = 4 - nf
      if (nb > 0) {
        b0 /= nb; b1 /= nb; b2 /= nb
      }
      if (nf === 0) {
        f0 = b0; f1 = b1; f2 = b2
      }
      let err = 0
      for (let k = 0; k < 4; k++) {
        const useF = g.cov[k] === 1
        const dr = quads[qb + k * 3] - (useF ? f0 : b0)
        const dg = quads[qb + k * 3 + 1] - (useF ? f1 : b1)
        const db = quads[qb + k * 3 + 2] - (useF ? f2 : b2)
        err += dr * dr * CHANNEL_W[0] + dg * dg * CHANNEL_W[1] + db * db * CHANNEL_W[2]
      }
      if (err < bestErr) {
        bestErr = err
        bgi = i
        bf0 = f0; bf1 = f1; bf2 = f2
        bg0 = b0; bg1 = b1; bg2 = b2
      }
    }
    glyph[c] = bgi
    fg[c * 3] = bf0; fg[c * 3 + 1] = bf1; fg[c * 3 + 2] = bf2
    bg[c * 3] = bg0; bg[c * 3 + 1] = bg1; bg[c * 3 + 2] = bg2
  }
  return { glyph, fg, bg }
}
